File: backend/app/schemas/fo.py
"""F&O analytics API schemas — Phase 4 slice 4.1."""

from __future__ import annotations

from datetime import date
from decimal import Decimal

from pydantic import BaseModel


class UnderlyingsOut(BaseModel):
    """F&O underlyings with option rows on the latest recorded day."""

    as_of: date | None
    symbols: list[str]


class ExpiryOut(BaseModel):
    expiry: date
    dte: int                  # calendar days from the chain's day to expiry


class ExpiriesOut(BaseModel):
    symbol: str
    as_of: date | None        # the recorded day these expiries are open on
    expiries: list[ExpiryOut]


class ChainLegOut(BaseModel):
    strike: Decimal
    option_type: str          # CE | PE
    oi: int
    volume: int
    ltp: Decimal | None
    # Populated only when ?greeks=true AND the quote inverts. None means
    # "not priced" — never zero, which would read as a real value on screen.
    iv: float | None = None           # annualized, Black-76 on the future
    delta: float | None = None
    gamma: float | None = None
    vega: float | None = None         # per 1.00 of IV (100 vol points)
    theta: float | None = None        # per year


class ChainOut(BaseModel):
    symbol: str
    expiry: date
    source: str               # eod | intraday
    spot: Decimal | None
    atm_strike: Decimal | None
    legs: list[ChainLegOut]
    # What the Greeks were priced off (null unless ?greeks=true resolved them).
    # Stated explicitly so the UI can never imply live Greeks off a stale chain.
    as_of: date | None = None         # the chain's own trading day
    fut_price: Decimal | None = None  # Black-76 forward
    dte: int | None = None            # calendar days from as_of to expiry


class PcrOut(BaseModel):
    pcr_oi: float | None
    pcr_volume: float | None
    total_ce_oi: int
    total_pe_oi: int


class BasisOut(BaseModel):
    fut_close: Decimal
    underlying_close: Decimal
    basis: Decimal
    basis_pct: float


class VixRegimeOut(BaseModel):
    current: Decimal
    percentile: float
    band: str                 # low | normal | high
    sample: int


class IvRankOut(BaseModel):
    symbol: str
    as_of: date
    current_iv: float
    rank: float
    percentile: float
    min_iv: float
    max_iv: float
    sample: int


class OptionLegOut(BaseModel):
    action: str          # sell | buy
    option_type: str     # CE | PE
    strike: Decimal
    premium: Decimal


class ExitPlanOut(BaseModel):
    take_profit_credit: Decimal
    stop_loss_amount: Decimal
    time_stop_dte: int


class SuggestionOut(BaseModel):
    structure: str       # bull_put | bear_call | iron_condor
    legs: list[OptionLegOut]
    net_credit: Decimal
    max_profit: Decimal
    max_loss: Decimal
    width: Decimal
    breakevens: list[Decimal]
    pop: float
    expectancy: Decimal
    margin_est: Decimal
    return_on_margin: float
    short_delta: float
    dte: int
    expiry: date
    exit_plan: ExitPlanOut
    rationale: str


class SuggestionsOut(BaseModel):
    symbol: str
    candidates: list[SuggestionOut]


class FoAnalyticsOut(BaseModel):
    symbol: str
    expiry: date
    source: str
    spot: Decimal | None
    atm_strike: Decimal | None
    pcr: PcrOut
    max_pain: Decimal | None
    basis: BasisOut | None
    vix: VixRegimeOut | None

File: frontend/src/lib/api/fo.ts
/**
 * F&O analytics client — Phase 4 backend (slices 4.1–4.3), Phase 5 UI.
 *
 * Money arrives as Decimal STRINGS and stays that way until a formatter in
 * lib/format.ts renders it; analytical values (IV, Greeks, PCR, POP) are
 * floats by design. `null` on a Greek means "not priced" — never 0.
 */

import { api } from './client'

export interface ChainLeg {
  strike: string
  option_type: 'CE' | 'PE'
  oi: number
  volume: number
  ltp: string | null
  // Present only when the chain was requested with greeks=true AND the quote
  // inverted. null = not priced (deep-ITM at intrinsic, no quote, no forward).
  iv: number | null
  delta: number | null
  gamma: number | null
  vega: number | null
  theta: number | null
}

export interface OptionChain {
  symbol: string
  expiry: string
  source: 'eod' | 'intraday'
  spot: string | null
  atm_strike: string | null
  legs: ChainLeg[]
  /** The chain's own trading day — what the Greeks were priced off. */
  as_of: string | null
  fut_price: string | null
  dte: number | null
}

export interface Pcr {
  pcr_oi: number | null
  pcr_volume: number | null
  total_ce_oi: number
  total_pe_oi: number
}

export interface Basis {
  fut_close: string
  underlying_close: string
  basis: string
  basis_pct: number
}

export interface VixRegime {
  current: string
  percentile: number
  band: 'low' | 'normal' | 'high'
  sample: number
}

export interface FoAnalytics {
  symbol: string
  expiry: string
  source: string
  spot: string | null
  atm_strike: string | null
  pcr: Pcr
  max_pain: string | null
  basis: Basis | null
  vix: VixRegime | null
}

export interface IvRank {
  symbol: string
  as_of: string
  current_iv: number
  rank: number
  percentile: number
  min_iv: number
  max_iv: number
  sample: number
}

export interface OptionLeg {
  action: 'sell' | 'buy'
  option_type: 'CE' | 'PE'
  strike: string
  premium: string
}

export interface ExitPlan {
  take_profit_credit: string
  stop_loss_amount: string
  time_stop_dte: number
}

export type SpreadStructure = 'bull_put' | 'bear_call' | 'iron_condor'

export interface SpreadCandidate {
  structure: SpreadStructure
  legs: OptionLeg[]
  net_credit: string
  max_profit: string
  max_loss: string
  width: string
  breakevens: string[]
  pop: number
  /** Report-only: risk-neutral expectancy is ~0 by construction. Never a gate. */
  expectancy: string
  margin_est: string
  return_on_margin: number
  short_delta: number
  dte: number
  expiry: string
  exit_plan: ExitPlan
  rationale: string
}

export interface SuggestionsResponse {
  symbol: string
  candidates: SpreadCandidate[]
}

export interface UnderlyingsResponse {
  as_of: string | null
  symbols: string[]
}

export interface ExpiryOption {
  expiry: string
  dte: number
}

export interface ExpiriesResponse {
  symbol: string
  as_of: string | null
  expiries: ExpiryOption[]
}

export const foApi = {
  getUnderlyings(token: string): Promise<UnderlyingsResponse> {
    return api.get<UnderlyingsResponse>('/fo/underlyings', token)
  },

  getExpiries(symbol: string, token: string): Promise<ExpiriesResponse> {
    return api.get<ExpiriesResponse>(`/fo/expiries?symbol=${encodeURIComponent(symbol)}`, token)
  },

  getChain(
    symbol: string,
    expiry: string,
    token: string,
    params: { source?: 'eod' | 'intraday'; strikes?: number; greeks?: boolean } = {},
  ): Promise<OptionChain> {
    const q = new URLSearchParams({ symbol, expiry })
    if (params.source) q.set('source', params.source)
    if (params.strikes != null) q.set('strikes', String(params.strikes))
    if (params.greeks) q.set('greeks', 'true')
    return api.get<OptionChain>(`/fo/chain?${q.toString()}`, token)
  },

  getAnalytics(
    symbol: string,
    expiry: string,
    token: string,
    params: { source?: 'eod' | 'intraday' } = {},
  ): Promise<FoAnalytics> {
    const q = new URLSearchParams({ symbol, expiry })
    if (params.source) q.set('source', params.source)
    return api.get<FoAnalytics>(`/fo/analytics?${q.toString()}`, token)
  },

  getIvRank(symbol: string, token: string): Promise<IvRank> {
    return api.get<IvRank>(`/fo/iv-rank?symbol=${encodeURIComponent(symbol)}`, token)
  },

  getSuggestions(symbol: string, token: string): Promise<SuggestionsResponse> {
    return api.get<SuggestionsResponse>(
      `/fo/suggestions?symbol=${encodeURIComponent(symbol)}`,
      token,
    )
  },
}
